// mediasoup/node/lib/fbs/data-consumer/message-notification.js
"use strict";
// automatically generated by the FlatBuffers compiler, do not modify
Object.defineProperty(exports, "__esModule", { value: true });
exports.MessageNotificationT = exports.MessageNotification = void 0;
/* eslint-disable @typescript-eslint/no-unused-vars, @typescript-eslint/no-explicit-any, @typescript-eslint/no-non-null-assertion */
const flatbuffers = require("flatbuffers");
class MessageNotification {
    bb = null;
    bb_pos = 0;
    __init(i, bb) {
        this.bb_pos = i;
        this.bb = bb;
        return this;
    }
    static getRootAsMessageNotification(bb, obj) {
        return (obj || new MessageNotification()).__init(bb.readInt32(bb.position()) + bb.position(), bb);
    }
    static getSizePrefixedRootAsMessageNotification(bb, obj) {
        bb.setPosition(bb.position() + flatbuffers.SIZE_PREFIX_LENGTH);
        return (obj || new MessageNotification()).__init(bb.readInt32(bb.position()) + bb.position(), bb);
    }
    ppid() {
        const offset = this.bb.__offset(this.bb_pos, 4);
        return offset ? this.bb.readUint32(this.bb_pos + offset) : 0;
    }
    data(index) {
        const offset = this.bb.__offset(this.bb_pos, 6);
        return offset ? this.bb.readUint8(this.bb.__vector(this.bb_pos + offset) + index) : 0;
    }
    dataLength() {
        const offset = this.bb.__offset(this.bb_pos, 6);
        return offset ? this.bb.__vector_len(this.bb_pos + offset) : 0;
    }
    dataArray() {
        const offset = this.bb.__offset(this.bb_pos, 6);
        return offset ? new Uint8Array(this.bb.bytes().buffer, this.bb.bytes().byteOffset + this.bb.__vector(this.bb_pos + offset), this.bb.__vector_len(this.bb_pos + offset)) : null;
    }
    static startMessageNotification(builder) {
        builder.startObject(2);
    }
    static addPpid(builder, ppid) {
        builder.addFieldInt32(0, ppid, 0);
    }
    static addData(builder, dataOffset) {
        builder.addFieldOffset(1, dataOffset, 0);
    }
    static createDataVector(builder, data) {
        builder.startVector(1, data.length, 1);
        for (let i = data.length - 1; i >= 0; i--) {
            builder.addInt8(data[i]);
        }
        return builder.endVector();
    }
    static startDataVector(builder, numElems) {
        builder.startVector(1, numElems, 1);
    }
    static endMessageNotification(builder) {
        const offset = builder.endObject();
        builder.requiredField(offset, 6); // data
        return offset;
    }
    static createMessageNotification(builder, ppid, dataOffset) {
        MessageNotification.startMessageNotification(builder);
        MessageNotification.addPpid(builder, ppid);
        MessageNotification.addData(builder, dataOffset);
        return MessageNotification.endMessageNotification(builder);
    }
    unpack() {
        return new MessageNotificationT(this.ppid(), this.bb.createScalarList(this.data.bind(this), this.dataLength()));
    }
    unpackTo(_o) {
        _o.ppid = this.ppid();
        _o.data = this.bb.createScalarList(this.data.bind(this), this.dataLength());
    }
}
exports.MessageNotification = MessageNotification;
class MessageNotificationT {
    ppid;
    data;
    constructor(ppid = 0, data = []) {
        this.ppid = ppid;
        this.data = data;
    }
    pack(builder) {
        const data = MessageNotification.createDataVector(builder, this.data);
        return MessageNotification.createMessageNotification(builder, this.ppid, data);
    }
}
exports.MessageNotificationT = MessageNotificationT;
