// mediasoup/node/lib/fbs/pipe-transport/pipe-transport-options.js
"use strict";
// automatically generated by the FlatBuffers compiler, do not modify
Object.defineProperty(exports, "__esModule", { value: true });
exports.PipeTransportOptionsT = exports.PipeTransportOptions = void 0;
/* eslint-disable @typescript-eslint/no-unused-vars, @typescript-eslint/no-explicit-any, @typescript-eslint/no-non-null-assertion */
const flatbuffers = require("flatbuffers");
const listen_info_1 = require("../../fbs/transport/listen-info");
const options_1 = require("../../fbs/transport/options");
class PipeTransportOptions {
    bb = null;
    bb_pos = 0;
    __init(i, bb) {
        this.bb_pos = i;
        this.bb = bb;
        return this;
    }
    static getRootAsPipeTransportOptions(bb, obj) {
        return (obj || new PipeTransportOptions()).__init(bb.readInt32(bb.position()) + bb.position(), bb);
    }
    static getSizePrefixedRootAsPipeTransportOptions(bb, obj) {
        bb.setPosition(bb.position() + flatbuffers.SIZE_PREFIX_LENGTH);
        return (obj || new PipeTransportOptions()).__init(bb.readInt32(bb.position()) + bb.position(), bb);
    }
    base(obj) {
        const offset = this.bb.__offset(this.bb_pos, 4);
        return offset ? (obj || new options_1.Options()).__init(this.bb.__indirect(this.bb_pos + offset), this.bb) : null;
    }
    listenInfo(obj) {
        const offset = this.bb.__offset(this.bb_pos, 6);
        return offset ? (obj || new listen_info_1.ListenInfo()).__init(this.bb.__indirect(this.bb_pos + offset), this.bb) : null;
    }
    enableRtx() {
        const offset = this.bb.__offset(this.bb_pos, 8);
        return offset ? !!this.bb.readInt8(this.bb_pos + offset) : false;
    }
    enableSrtp() {
        const offset = this.bb.__offset(this.bb_pos, 10);
        return offset ? !!this.bb.readInt8(this.bb_pos + offset) : false;
    }
    static startPipeTransportOptions(builder) {
        builder.startObject(4);
    }
    static addBase(builder, baseOffset) {
        builder.addFieldOffset(0, baseOffset, 0);
    }
    static addListenInfo(builder, listenInfoOffset) {
        builder.addFieldOffset(1, listenInfoOffset, 0);
    }
    static addEnableRtx(builder, enableRtx) {
        builder.addFieldInt8(2, +enableRtx, +false);
    }
    static addEnableSrtp(builder, enableSrtp) {
        builder.addFieldInt8(3, +enableSrtp, +false);
    }
    static endPipeTransportOptions(builder) {
        const offset = builder.endObject();
        builder.requiredField(offset, 4); // base
        builder.requiredField(offset, 6); // listen_info
        return offset;
    }
    unpack() {
        return new PipeTransportOptionsT((this.base() !== null ? this.base().unpack() : null), (this.listenInfo() !== null ? this.listenInfo().unpack() : null), this.enableRtx(), this.enableSrtp());
    }
    unpackTo(_o) {
        _o.base = (this.base() !== null ? this.base().unpack() : null);
        _o.listenInfo = (this.listenInfo() !== null ? this.listenInfo().unpack() : null);
        _o.enableRtx = this.enableRtx();
        _o.enableSrtp = this.enableSrtp();
    }
}
exports.PipeTransportOptions = PipeTransportOptions;
class PipeTransportOptionsT {
    base;
    listenInfo;
    enableRtx;
    enableSrtp;
    constructor(base = null, listenInfo = null, enableRtx = false, enableSrtp = false) {
        this.base = base;
        this.listenInfo = listenInfo;
        this.enableRtx = enableRtx;
        this.enableSrtp = enableSrtp;
    }
    pack(builder) {
        const base = (this.base !== null ? this.base.pack(builder) : 0);
        const listenInfo = (this.listenInfo !== null ? this.listenInfo.pack(builder) : 0);
        PipeTransportOptions.startPipeTransportOptions(builder);
        PipeTransportOptions.addBase(builder, base);
        PipeTransportOptions.addListenInfo(builder, listenInfo);
        PipeTransportOptions.addEnableRtx(builder, this.enableRtx);
        PipeTransportOptions.addEnableSrtp(builder, this.enableSrtp);
        return PipeTransportOptions.endPipeTransportOptions(builder);
    }
}
exports.PipeTransportOptionsT = PipeTransportOptionsT;
