// mediasoup/node/lib/fbs/response/body.js
"use strict";
// automatically generated by the FlatBuffers compiler, do not modify
Object.defineProperty(exports, "__esModule", { value: true });
exports.Body = void 0;
exports.unionToBody = unionToBody;
exports.unionListToBody = unionListToBody;
/* eslint-disable @typescript-eslint/no-unused-vars, @typescript-eslint/no-explicit-any, @typescript-eslint/no-non-null-assertion */
const dump_response_1 = require("../../fbs/consumer/dump-response");
const get_stats_response_1 = require("../../fbs/consumer/get-stats-response");
const set_preferred_layers_response_1 = require("../../fbs/consumer/set-preferred-layers-response");
const set_priority_response_1 = require("../../fbs/consumer/set-priority-response");
const add_subchannel_response_1 = require("../../fbs/data-consumer/add-subchannel-response");
const dump_response_2 = require("../../fbs/data-consumer/dump-response");
const get_buffered_amount_response_1 = require("../../fbs/data-consumer/get-buffered-amount-response");
const get_stats_response_2 = require("../../fbs/data-consumer/get-stats-response");
const remove_subchannel_response_1 = require("../../fbs/data-consumer/remove-subchannel-response");
const set_subchannels_response_1 = require("../../fbs/data-consumer/set-subchannels-response");
const dump_response_3 = require("../../fbs/data-producer/dump-response");
const get_stats_response_3 = require("../../fbs/data-producer/get-stats-response");
const dump_response_4 = require("../../fbs/direct-transport/dump-response");
const get_stats_response_4 = require("../../fbs/direct-transport/get-stats-response");
const connect_response_1 = require("../../fbs/pipe-transport/connect-response");
const dump_response_5 = require("../../fbs/pipe-transport/dump-response");
const get_stats_response_5 = require("../../fbs/pipe-transport/get-stats-response");
const connect_response_2 = require("../../fbs/plain-transport/connect-response");
const dump_response_6 = require("../../fbs/plain-transport/dump-response");
const get_stats_response_6 = require("../../fbs/plain-transport/get-stats-response");
const dump_response_7 = require("../../fbs/producer/dump-response");
const get_stats_response_7 = require("../../fbs/producer/get-stats-response");
const dump_response_8 = require("../../fbs/router/dump-response");
const consume_response_1 = require("../../fbs/transport/consume-response");
const produce_response_1 = require("../../fbs/transport/produce-response");
const restart_ice_response_1 = require("../../fbs/transport/restart-ice-response");
const dump_response_9 = require("../../fbs/web-rtc-server/dump-response");
const connect_response_3 = require("../../fbs/web-rtc-transport/connect-response");
const dump_response_10 = require("../../fbs/web-rtc-transport/dump-response");
const get_stats_response_8 = require("../../fbs/web-rtc-transport/get-stats-response");
const dump_response_11 = require("../../fbs/worker/dump-response");
const resource_usage_response_1 = require("../../fbs/worker/resource-usage-response");
var Body;
(function (Body) {
    Body[Body["NONE"] = 0] = "NONE";
    Body[Body["Worker_DumpResponse"] = 1] = "Worker_DumpResponse";
    Body[Body["Worker_ResourceUsageResponse"] = 2] = "Worker_ResourceUsageResponse";
    Body[Body["WebRtcServer_DumpResponse"] = 3] = "WebRtcServer_DumpResponse";
    Body[Body["Router_DumpResponse"] = 4] = "Router_DumpResponse";
    Body[Body["Transport_ProduceResponse"] = 5] = "Transport_ProduceResponse";
    Body[Body["Transport_ConsumeResponse"] = 6] = "Transport_ConsumeResponse";
    Body[Body["Transport_RestartIceResponse"] = 7] = "Transport_RestartIceResponse";
    Body[Body["PlainTransport_ConnectResponse"] = 8] = "PlainTransport_ConnectResponse";
    Body[Body["PlainTransport_DumpResponse"] = 9] = "PlainTransport_DumpResponse";
    Body[Body["PlainTransport_GetStatsResponse"] = 10] = "PlainTransport_GetStatsResponse";
    Body[Body["PipeTransport_ConnectResponse"] = 11] = "PipeTransport_ConnectResponse";
    Body[Body["PipeTransport_DumpResponse"] = 12] = "PipeTransport_DumpResponse";
    Body[Body["PipeTransport_GetStatsResponse"] = 13] = "PipeTransport_GetStatsResponse";
    Body[Body["DirectTransport_DumpResponse"] = 14] = "DirectTransport_DumpResponse";
    Body[Body["DirectTransport_GetStatsResponse"] = 15] = "DirectTransport_GetStatsResponse";
    Body[Body["WebRtcTransport_ConnectResponse"] = 16] = "WebRtcTransport_ConnectResponse";
    Body[Body["WebRtcTransport_DumpResponse"] = 17] = "WebRtcTransport_DumpResponse";
    Body[Body["WebRtcTransport_GetStatsResponse"] = 18] = "WebRtcTransport_GetStatsResponse";
    Body[Body["Producer_DumpResponse"] = 19] = "Producer_DumpResponse";
    Body[Body["Producer_GetStatsResponse"] = 20] = "Producer_GetStatsResponse";
    Body[Body["Consumer_DumpResponse"] = 21] = "Consumer_DumpResponse";
    Body[Body["Consumer_GetStatsResponse"] = 22] = "Consumer_GetStatsResponse";
    Body[Body["Consumer_SetPreferredLayersResponse"] = 23] = "Consumer_SetPreferredLayersResponse";
    Body[Body["Consumer_SetPriorityResponse"] = 24] = "Consumer_SetPriorityResponse";
    Body[Body["DataProducer_DumpResponse"] = 25] = "DataProducer_DumpResponse";
    Body[Body["DataProducer_GetStatsResponse"] = 26] = "DataProducer_GetStatsResponse";
    Body[Body["DataConsumer_GetBufferedAmountResponse"] = 27] = "DataConsumer_GetBufferedAmountResponse";
    Body[Body["DataConsumer_DumpResponse"] = 28] = "DataConsumer_DumpResponse";
    Body[Body["DataConsumer_GetStatsResponse"] = 29] = "DataConsumer_GetStatsResponse";
    Body[Body["DataConsumer_SetSubchannelsResponse"] = 30] = "DataConsumer_SetSubchannelsResponse";
    Body[Body["DataConsumer_AddSubchannelResponse"] = 31] = "DataConsumer_AddSubchannelResponse";
    Body[Body["DataConsumer_RemoveSubchannelResponse"] = 32] = "DataConsumer_RemoveSubchannelResponse";
})(Body || (exports.Body = Body = {}));
function unionToBody(type, accessor) {
    switch (Body[type]) {
        case 'NONE': return null;
        case 'Worker_DumpResponse': return accessor(new dump_response_11.DumpResponse());
        case 'Worker_ResourceUsageResponse': return accessor(new resource_usage_response_1.ResourceUsageResponse());
        case 'WebRtcServer_DumpResponse': return accessor(new dump_response_9.DumpResponse());
        case 'Router_DumpResponse': return accessor(new dump_response_8.DumpResponse());
        case 'Transport_ProduceResponse': return accessor(new produce_response_1.ProduceResponse());
        case 'Transport_ConsumeResponse': return accessor(new consume_response_1.ConsumeResponse());
        case 'Transport_RestartIceResponse': return accessor(new restart_ice_response_1.RestartIceResponse());
        case 'PlainTransport_ConnectResponse': return accessor(new connect_response_2.ConnectResponse());
        case 'PlainTransport_DumpResponse': return accessor(new dump_response_6.DumpResponse());
        case 'PlainTransport_GetStatsResponse': return accessor(new get_stats_response_6.GetStatsResponse());
        case 'PipeTransport_ConnectResponse': return accessor(new connect_response_1.ConnectResponse());
        case 'PipeTransport_DumpResponse': return accessor(new dump_response_5.DumpResponse());
        case 'PipeTransport_GetStatsResponse': return accessor(new get_stats_response_5.GetStatsResponse());
        case 'DirectTransport_DumpResponse': return accessor(new dump_response_4.DumpResponse());
        case 'DirectTransport_GetStatsResponse': return accessor(new get_stats_response_4.GetStatsResponse());
        case 'WebRtcTransport_ConnectResponse': return accessor(new connect_response_3.ConnectResponse());
        case 'WebRtcTransport_DumpResponse': return accessor(new dump_response_10.DumpResponse());
        case 'WebRtcTransport_GetStatsResponse': return accessor(new get_stats_response_8.GetStatsResponse());
        case 'Producer_DumpResponse': return accessor(new dump_response_7.DumpResponse());
        case 'Producer_GetStatsResponse': return accessor(new get_stats_response_7.GetStatsResponse());
        case 'Consumer_DumpResponse': return accessor(new dump_response_1.DumpResponse());
        case 'Consumer_GetStatsResponse': return accessor(new get_stats_response_1.GetStatsResponse());
        case 'Consumer_SetPreferredLayersResponse': return accessor(new set_preferred_layers_response_1.SetPreferredLayersResponse());
        case 'Consumer_SetPriorityResponse': return accessor(new set_priority_response_1.SetPriorityResponse());
        case 'DataProducer_DumpResponse': return accessor(new dump_response_3.DumpResponse());
        case 'DataProducer_GetStatsResponse': return accessor(new get_stats_response_3.GetStatsResponse());
        case 'DataConsumer_GetBufferedAmountResponse': return accessor(new get_buffered_amount_response_1.GetBufferedAmountResponse());
        case 'DataConsumer_DumpResponse': return accessor(new dump_response_2.DumpResponse());
        case 'DataConsumer_GetStatsResponse': return accessor(new get_stats_response_2.GetStatsResponse());
        case 'DataConsumer_SetSubchannelsResponse': return accessor(new set_subchannels_response_1.SetSubchannelsResponse());
        case 'DataConsumer_AddSubchannelResponse': return accessor(new add_subchannel_response_1.AddSubchannelResponse());
        case 'DataConsumer_RemoveSubchannelResponse': return accessor(new remove_subchannel_response_1.RemoveSubchannelResponse());
        default: return null;
    }
}
function unionListToBody(type, accessor, index) {
    switch (Body[type]) {
        case 'NONE': return null;
        case 'Worker_DumpResponse': return accessor(index, new dump_response_11.DumpResponse());
        case 'Worker_ResourceUsageResponse': return accessor(index, new resource_usage_response_1.ResourceUsageResponse());
        case 'WebRtcServer_DumpResponse': return accessor(index, new dump_response_9.DumpResponse());
        case 'Router_DumpResponse': return accessor(index, new dump_response_8.DumpResponse());
        case 'Transport_ProduceResponse': return accessor(index, new produce_response_1.ProduceResponse());
        case 'Transport_ConsumeResponse': return accessor(index, new consume_response_1.ConsumeResponse());
        case 'Transport_RestartIceResponse': return accessor(index, new restart_ice_response_1.RestartIceResponse());
        case 'PlainTransport_ConnectResponse': return accessor(index, new connect_response_2.ConnectResponse());
        case 'PlainTransport_DumpResponse': return accessor(index, new dump_response_6.DumpResponse());
        case 'PlainTransport_GetStatsResponse': return accessor(index, new get_stats_response_6.GetStatsResponse());
        case 'PipeTransport_ConnectResponse': return accessor(index, new connect_response_1.ConnectResponse());
        case 'PipeTransport_DumpResponse': return accessor(index, new dump_response_5.DumpResponse());
        case 'PipeTransport_GetStatsResponse': return accessor(index, new get_stats_response_5.GetStatsResponse());
        case 'DirectTransport_DumpResponse': return accessor(index, new dump_response_4.DumpResponse());
        case 'DirectTransport_GetStatsResponse': return accessor(index, new get_stats_response_4.GetStatsResponse());
        case 'WebRtcTransport_ConnectResponse': return accessor(index, new connect_response_3.ConnectResponse());
        case 'WebRtcTransport_DumpResponse': return accessor(index, new dump_response_10.DumpResponse());
        case 'WebRtcTransport_GetStatsResponse': return accessor(index, new get_stats_response_8.GetStatsResponse());
        case 'Producer_DumpResponse': return accessor(index, new dump_response_7.DumpResponse());
        case 'Producer_GetStatsResponse': return accessor(index, new get_stats_response_7.GetStatsResponse());
        case 'Consumer_DumpResponse': return accessor(index, new dump_response_1.DumpResponse());
        case 'Consumer_GetStatsResponse': return accessor(index, new get_stats_response_1.GetStatsResponse());
        case 'Consumer_SetPreferredLayersResponse': return accessor(index, new set_preferred_layers_response_1.SetPreferredLayersResponse());
        case 'Consumer_SetPriorityResponse': return accessor(index, new set_priority_response_1.SetPriorityResponse());
        case 'DataProducer_DumpResponse': return accessor(index, new dump_response_3.DumpResponse());
        case 'DataProducer_GetStatsResponse': return accessor(index, new get_stats_response_3.GetStatsResponse());
        case 'DataConsumer_GetBufferedAmountResponse': return accessor(index, new get_buffered_amount_response_1.GetBufferedAmountResponse());
        case 'DataConsumer_DumpResponse': return accessor(index, new dump_response_2.DumpResponse());
        case 'DataConsumer_GetStatsResponse': return accessor(index, new get_stats_response_2.GetStatsResponse());
        case 'DataConsumer_SetSubchannelsResponse': return accessor(index, new set_subchannels_response_1.SetSubchannelsResponse());
        case 'DataConsumer_AddSubchannelResponse': return accessor(index, new add_subchannel_response_1.AddSubchannelResponse());
        case 'DataConsumer_RemoveSubchannelResponse': return accessor(index, new remove_subchannel_response_1.RemoveSubchannelResponse());
        default: return null;
    }
}
