// mediasoup/node/lib/fbs/web-rtc-transport/listen.js
"use strict";
// automatically generated by the FlatBuffers compiler, do not modify
Object.defineProperty(exports, "__esModule", { value: true });
exports.Listen = void 0;
exports.unionToListen = unionToListen;
exports.unionListToListen = unionListToListen;
/* eslint-disable @typescript-eslint/no-unused-vars, @typescript-eslint/no-explicit-any, @typescript-eslint/no-non-null-assertion */
const listen_individual_1 = require("../../fbs/web-rtc-transport/listen-individual");
const listen_server_1 = require("../../fbs/web-rtc-transport/listen-server");
var Listen;
(function (Listen) {
    Listen[Listen["NONE"] = 0] = "NONE";
    Listen[Listen["ListenIndividual"] = 1] = "ListenIndividual";
    Listen[Listen["ListenServer"] = 2] = "ListenServer";
})(Listen || (exports.Listen = Listen = {}));
function unionToListen(type, accessor) {
    switch (Listen[type]) {
        case 'NONE': return null;
        case 'ListenIndividual': return accessor(new listen_individual_1.ListenIndividual());
        case 'ListenServer': return accessor(new listen_server_1.ListenServer());
        default: return null;
    }
}
function unionListToListen(type, accessor, index) {
    switch (Listen[type]) {
        case 'NONE': return null;
        case 'ListenIndividual': return accessor(index, new listen_individual_1.ListenIndividual());
        case 'ListenServer': return accessor(index, new listen_server_1.ListenServer());
        default: return null;
    }
}
