// mediasoup/node/lib/fbs/rtp-stream.js
"use strict";
// automatically generated by the FlatBuffers compiler, do not modify
Object.defineProperty(exports, "__esModule", { value: true });
exports.StatsData = exports.StatsT = exports.Stats = exports.SendStatsT = exports.SendStats = exports.RecvStatsT = exports.RecvStats = exports.ParamsT = exports.Params = exports.DumpT = exports.Dump = exports.BitrateByLayerT = exports.BitrateByLayer = exports.BaseStatsT = exports.BaseStats = void 0;
/* eslint-disable @typescript-eslint/no-unused-vars, @typescript-eslint/no-explicit-any, @typescript-eslint/no-non-null-assertion */
var base_stats_1 = require("./rtp-stream/base-stats");
Object.defineProperty(exports, "BaseStats", { enumerable: true, get: function () { return base_stats_1.BaseStats; } });
Object.defineProperty(exports, "BaseStatsT", { enumerable: true, get: function () { return base_stats_1.BaseStatsT; } });
var bitrate_by_layer_1 = require("./rtp-stream/bitrate-by-layer");
Object.defineProperty(exports, "BitrateByLayer", { enumerable: true, get: function () { return bitrate_by_layer_1.BitrateByLayer; } });
Object.defineProperty(exports, "BitrateByLayerT", { enumerable: true, get: function () { return bitrate_by_layer_1.BitrateByLayerT; } });
var dump_1 = require("./rtp-stream/dump");
Object.defineProperty(exports, "Dump", { enumerable: true, get: function () { return dump_1.Dump; } });
Object.defineProperty(exports, "DumpT", { enumerable: true, get: function () { return dump_1.DumpT; } });
var params_1 = require("./rtp-stream/params");
Object.defineProperty(exports, "Params", { enumerable: true, get: function () { return params_1.Params; } });
Object.defineProperty(exports, "ParamsT", { enumerable: true, get: function () { return params_1.ParamsT; } });
var recv_stats_1 = require("./rtp-stream/recv-stats");
Object.defineProperty(exports, "RecvStats", { enumerable: true, get: function () { return recv_stats_1.RecvStats; } });
Object.defineProperty(exports, "RecvStatsT", { enumerable: true, get: function () { return recv_stats_1.RecvStatsT; } });
var send_stats_1 = require("./rtp-stream/send-stats");
Object.defineProperty(exports, "SendStats", { enumerable: true, get: function () { return send_stats_1.SendStats; } });
Object.defineProperty(exports, "SendStatsT", { enumerable: true, get: function () { return send_stats_1.SendStatsT; } });
var stats_1 = require("./rtp-stream/stats");
Object.defineProperty(exports, "Stats", { enumerable: true, get: function () { return stats_1.Stats; } });
Object.defineProperty(exports, "StatsT", { enumerable: true, get: function () { return stats_1.StatsT; } });
var stats_data_1 = require("./rtp-stream/stats-data");
Object.defineProperty(exports, "StatsData", { enumerable: true, get: function () { return stats_data_1.StatsData; } });
